### ISSUE TITLE
refactor: inline and extract variables

### ISSUE THREAD

**Issue by Trass3r** [clangd/clangd#446]

**Comment by kirillbobyrev** [comment]
I can't find anything about it in the documentation which we probably should fix but we do have variable and method extracting actions.

We should update https://clangd.llvm.org/features.html#refactoring.

https://github.com/llvm/llvm-project/blob/master/clang-tools-extra/clangd/refactor/tweaks/ExtractVariable.cpp
https://github.com/llvm/llvm-project/blob/master/clang-tools-extra/clangd/refactor/tweaks/ExtractFunction.cpp

**Comment by Trass3r** [comment]
Ok I quickly tried selecting some code in VSCode but but Refactor didn't offer anything.

**Comment by kirillbobyrev** [comment]
There could be several reasons:

* Clangd you're using is too old. I think this feature was added in... 10.0, maybe 9.0? Not sure about 9.0, but I've used 10.0, so it's there for sure.
* The code you're selecting is too complex for the refactoring actions.

As far as I understand it's pretty basic and quite limited (there is a lot to figure out - passing the right parameters and getting dependencies right, name resolution and ADL, and so on).

It would be helpful if you provided an example of what is failing, otherwise, I can't really understand what's going wrong and what you expect to work.

![Extract-Prompt](https://user-images.githubusercontent.com/3352968/86512874-59955580-be0e-11ea-9e5b-e2028c2c3f14.png)
![Extract-Prompt-Text](https://user-images.githubusercontent.com/3352968/86512876-5c904600-be0e-11ea-9168-b13e00e84176.png)
![Extract-Result](https://user-images.githubusercontent.com/3352968/86512879-5dc17300-be0e-11ea-822c-9df54da241a7.png)

**Comment by Trass3r** [comment]
`clangd version 11.0.0-++20200701052636+b2aa546b074-1~exp1~20200701153313.89`
I tried your example, no refactorings offered.

**Comment by kirillbobyrev** [comment]
Hm, that is unexpected, I need to try this with a new version of Clangd but I wouldn't expect regressions here. Any chance your setup is somehow interacting with this feature? Do you get refactorings for other languages?

**Comment by Trass3r** [comment]
Yes that refactoring works for TypeScript for example.

**Comment by kirillbobyrev** [comment]
Hm, this is interesting. I've checked both VSCode and coc-clangd, both Clangd 10.0 and latest Clangd and still see this refactoring suggestion from the picture. Can't figure out why it isn't shown to you right now.

Making it more robust and advanced is certainly a good thing, though, in its current form it's somewhat limited.

**Comment by Trass3r** [comment]
Could clangd parameters affect this? I'm using `--background-index --pch-storage=memory --suggest-missing-includes --debug-origin --cross-file-rename --all-scopes-completion --clang-tidy --pretty`.

**Comment by kirillbobyrev** [comment]
I don't think so. Also, most flags you have are on by default (background index, clang-tidy, suggest missing includes, cross file rename, all scopes completion). PCH storage and debug origin shouldn't really affect this and I've confirmed that with my editor. Do you not have a light bulb upon selecting the code region from the example at all?

**Comment by Trass3r** [comment]
I tried again, your example works now, odd. I did update clangd and use https://github.com/clangd/vscode-clangd/pull/54 now but I wouldn't expect that to affect the behavior.

**Comment by kirillbobyrev** [comment]
Hmm, this is really interesting, you're right, I wouldn't expect this to be the root cause. Maybe the version you used was a temporary regression/broken in some way that is fixed now.

Thanks for the update!
